src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2908,6 +2908,15 @@ module Crystal
       node.elements.each &.add_observer(node)
       node.program = @program
       node.update
+
+      node.elements.each do |element|
+        if element.is_a?(Splat) && (type = element.type?)
+          unless type.is_a?(TupleInstanceType)
+            node.raise "argument to splat must be a tuple, not #{type}"
+          end
+        end
+      end
```

Comment:
I am starting to think if we could do the following instead, and remove similar checks in other places:

```crystal
module Crystal
  class MainVisitor
    def end_visit(node : Splat)
      node.bind_to node.exp

      if (type = node.exp.type?) && !type.is_a?(TupleInstanceType)
        node.raise "argument to splat must be a tuple, not #{type}"
      end
    end
  end
end
```

Running all the semantic tests locally reveals exactly two failing test cases which only affect error messages. The first one is a slight wording change:

```crystal
it "errors if splatting a non-tuple (#9853)" do
  assert_error %(
    Array(*Int32)
    ),
    "argument to splat must be a tuple type, not Int32"
  # got: "argument to splat must be a tuple, not Int32"
end
```

The second is

```crystal
it "errors if splatting union" do
  assert_error %(
    a = {1} || {1, 2}
    foo *a
    ),
    "not yet supported"
  # got: "argument to splat must be a tuple, not (Tuple(Int32) | Tuple(Int32, Int32))"
end
```

This is also probably acceptable since Crystal already merges unions of tuples of the same arity (not that I agree with it), so the unions that aren't reduced to tuples are precisely ones containing tuples of different arities.

Reply:
This can probably be done in CleanUpTransformer (or something like that) where all types are known